(Feat) Posting Journal Editing

jniles opened the pull request:
This PR adds the basic format for creating ui-grid editors by implementing a datepicker editor.  The editor is found in the directive `bhGridDateEditor.js`.  For technical details, see the code!

**New Methods**
Additionally, this commit introduces two new functions: `util.before()` and `util.after()` for calling functions before or after specific class/object method calls.  They can be used like this:

``` js
var klass = { method : p => console.log('Called with p:', p) };
util.before(klass, 'method', p => {
  console.log('I got injected before!  I found parameter:', p);
});

klass.method(2);
// => I got injected before!  I found parameter: 2
// => Called with p: 2

```

`util.after()` is similar, except the call is made after the method call is executed.

**New Features**
This PR introduces a loading indicator to the Posting Journal.  It leverages the `bhGridLoadingIndicator` directive.

**Tests**
This PR also introduces unit tests for:
1. `util.once()` introduced in #598.
2. `util.before()` introduced here.
3. `util.after()` introduced here.

**TODO**
This PR only implements an editor - none of the editing logic.  Here was what remains to be done:
1. ~~Limit edits to a single transaction at a time.~~
2. ~~Implement a "Save" button to validate and terminate the edits on a transaction.~~
3. When trying to save an `$invalid` input value either:
   1. Replace the edits with the original data.
   2. Do not allow the user to quit editing until the inputs are `$valid`
4. ~~Highlight the transaction currently being edited.~~

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

Commit:
test(util): add AOP methods + test to util service

This commit adds aspect-oriented programming methods to the util service
for better readability of Grid Services.  It adds unit tests that show
functionality but do little to assure correctness.  These should be
expanded in the future as more use-cases for AOP become evident.

Most of the Grid Services use `onRegisterApi()` to bind the grid's API
to the service instance and perform specific actions before or after the
API becomes available.  This is the perfect use-case for aspect oriented
programming, so some naive methods have been implemented in the `util`
service library.

Additionally, this commit renames files without changing their contexts
for better clarity.

Commit:
chore(grid): use util.after() for API registry

This commit replaces all instances of manual replacement of
onRegisterApi() with util.after() calls for better predictability of the
order of callbacks executed by the onRegisterApi() call.

Commit:
chore(journal): add loading indicator to grid

This commit adds a loading indicator to the grid for a nicer loading
animation when there are many records to fetch.

Commit:
feat(grid): use ui-grid-datepicker as date editor

This commit uses a pre-built ui-datepicker as the date editor for the
ui-grid.  This allows the grid to edit dates with a datepicker.

Commit:
feat(bhGridDateEditor): implement ui-grid datepicker

This commit implements a ui-grid datepicker directive for transaction
date on the posting journal (although it can be used anywhere).  The
directive still needs lots of work to be fully functional, including
validation and super user edit permission.

Commit:
deps(pkg): version bump most greenkeeper packages

This commit simply bumps the versions of many packages to clean out
greenkeeper PRs.

Commit:
feat(journal): only edit editable transactions

This commit implements highlighting and edit mode for transaction groups
in the journal ui-grid.  It also ensures that only editable transactions
(those marked with _editing) are allowed to receive edits.

Finally, an editing flag is visible outside of the ui-grid to show that
there are currently transactions open for editing.

Commit:
feat(journal): allow grid editing on journal

This commit implements editing on a transaction level.  The workflow is
like this:

 1. The user clicks the _edit_ button in the action column of the grid.
 This puts the transaction (and all internal rows) in _editing mode.
 2. The user is then able to navigate using the cellNav plugin to edit
 consecutive cells.
 3. When the user is finished, they are able to turn editing off via
 the 'Save' button in the top right corner of the transaction header.

There are still the following TODOs:
 1. Disallow cellNav on certain columns (action column).
 2. Automatically focus the first editable column in the transaction
 when the "Edit" button is clicked.
 3. Make transaction constants for the following states:
  1. Edit
  2. Highlight
  3. Invalid

Commit:
chore(journal): turn off inline filtering default

This commit turns off the inline filtering on the posting journal by
default to give more surface area for working on the grid.  It should be
enabled in a configuration option as needed.

Commit:
chore(journal): disallow cell focus on some cols

This commit disallows cell focus on some non-editable columns to ease
user experience.  Those columns are:
 1. Trans_id
 2. Action

Commit:
refactor(journal): use constants for grid flags

This commit updates the Transaction service and all associated templates
to use global variables for EDIT, INVALID, and HIGHLIGHT states.

Commit:
fix(journal): ensure the saved cells unfocus grid

This commit ensures that the grid unfocuses transactions on save for a
better UX.

sfount commented on client/src/partials/journal/journal.html:
```diff
@@ -2,7 +2,21 @@
+
+        <span class="badge badge-primary text-uppercase" ng-show="JournalCtrl.transactions.isEditing()">
+          <i class="fa fa-street-view"></i>
+          Edit Mode
```
This should eventually be translated - the feature is still in early development so this is just a reminder comment.

sfount commented on client/src/partials/journal/journal.html:
```diff
@@ -2,7 +2,21 @@
+
+        <span class="badge badge-warning text-uppercase" ng-show="JournalCtrl.transactions._highlights.length > 0">
+          {{ JournalCtrl.transactions._highlights.length }}
+          <span ng-if="JournalCtrl.transactions._highlights.length == 1">Highlight</span>
```
Are highlights a feature that will be implemented in the future or that were used during development/ debugging? 
I can't see any way that this state is ever reached in this PR.

jniles replied:
They were mostly for prototyping, although they may be useful in highlighting transactions is coming from a previous state.  Something like:

`#/journal/:uuid` --> highlight transaction with record UUID `:uuid`.  This is unimplemented.

But I'll take this out for the time being.  They are easy to add back in.

sfount commented on client/src/js/services/grid/GridSorting.js:
```diff
@@ -77,23 +79,14 @@ function GridSortingService() {
-    // // Register for the Grid API
-    gridOptions.onRegisterApi = function onRegisterApi(api) {
+    // register for the grid API
+    util.after(gridOptions, 'onRegisterApi', function onRegisterApi(api) {
```
Cool 👍

Commit:
refactor(journal): rm highlighting

This commit removes the highlighting feature for the time being, until
there is an actual use case for it.

It also temporarily skips the reference lock test until transaction
restarts are implemented.  See #620.
